pkg/controllers/podopslifecycle/podopslifecycle_controller.go

```diff
@@ -193,19 +194,94 @@ func (r *ReconcilePodOpsLifecycle) addServiceAvailable(pod *corev1.Pod) (bool, e
 		return false, nil
 	}
 
-	satisfied, _, err := controllerutils.SatisfyExpectedFinalizers(pod) // whether all expected finalizers are satisfied
-	if err != nil || !satisfied {
+	satisfied, notSatisfiedFinalizers, err := controllerutils.SatisfyExpectedFinalizers(pod) // whether all expected finalizers are satisfied
+	if err != nil {
 		return false, err
 	}
 
+	if !satisfied {
+		notDirtyExist, err := r.removeDirtyExpectedFinalizer(pod, notSatisfiedFinalizers)
+		if err != nil {
+			return false, err
+		}
+		if notDirtyExist {
+			return false, nil
+		}
+		// all not satisfied expected finalizers are dirty, so actually the pod satisfied expected finalizer now
+	}
+
 	if !controllerutils.IsPodReady(pod) {
 		return false, nil
 	}
 
-	labels := map[string]string{
+	podLabelsToAdd := map[string]string{
 		v1alpha1.PodServiceAvailableLabel: strconv.FormatInt(time.Now().Unix(), 10),
 	}
-	return true, r.addLabels(context.Background(), pod, labels)
+	return true, r.addLabels(context.Background(), pod, podLabelsToAdd)
+}
+
+func (r *ReconcilePodOpsLifecycle) removeDirtyExpectedFinalizer(pod *corev1.Pod, notSatisfiedFinalizers map[string]string) (bool, error) {
+	var notDirtyExist bool
```

Comment:
isDirty is more easy to understand.

Reply:
allDirty might be better
!allDirty equals notDirtyExist